src/sparseml/pytorch/sparsification/pruning/mask_creator.py

```diff
@@ -0,0 +1,424 @@
+            unmasked) calculated from the tensors such that the desired number of zeros
+            matches the sparsity and all values mapped to the same group have the same
+            value
+        """
+        sparsity = target
+        grouped_tensors = self._group_tensors(tensors)
+        grouped_masks = super().create_sparsity_masks(
+            grouped_tensors, sparsity, global_sparsity
+        )
+        masks = [
+            self._map_mask_to_tensor(grouped_mask, tensor.shape, idx)
+            for idx, (grouped_mask, tensor) in enumerate(zip(grouped_masks, tensors))
+        ]
+
+        return masks
+
+    def _group_tensors(self, tensors: List[Tensor]) -> List[Tensor]:
+        return [self.group_tensor(tensor) for tensor in tensors]
+
+
+class FourBlockMaskCreator(GroupedPruningMaskCreator):
+    """
+    semi-structured sparsity mask creator that groups sparsity blocks in groups of four
+    along the input-channel dimension (assumed to be dimension 1 for pytorch)
+
+    Equivalent to BlockPruningMaskCreator([1, 4]) without restrictions on number
+    of dimensions, or divisibility
+
+    :param grouping_fn_name: The name of the torch grouping function to reduce
+        dimensions by
+    """
+
+    def __init__(
+        self,
+        grouping_fn_name: str = "mean",
+    ):
+        self._grouping_fn_name = grouping_fn_name
+
+    def group_tensor(self, tensor: Tensor) -> Tensor:
+        """
+        :param tensor: The tensor to transform
+        :return: The mean values of the tensor grouped by blocks of shape
+            self._block_shape
+        """
+        if tensor.dim() > 2:
+            # permute input channel dim to last dimension
+            permute_val = list(range(tensor.dim()))
+            del permute_val[1]
+            permute_val.append(1)
+            tensor = tensor.permute(*permute_val)
+
+        remainder = tensor.size(-1) % 4
+        if remainder != 0:
+            # pad with zeros to make masks add to 4
+            pad_num = 4 - remainder
+            padded_tensor = torch.zeros(
+                *tensor.shape[:-1],
+                tensor.size(-1) + pad_num,
+                device=tensor.device,
+                dtype=tensor.dtype,
+            )
+            padded_tensor[..., :-pad_num] = tensor
+            padded_tensor[..., -pad_num:] = torch.mean(
+                # mean of remainder input channel dims
+                tensor[..., -remainder:],
+                dim=-1,
+                keepdim=True,
+            )
+            tensor = padded_tensor
+
+        blocked_tensor = tensor.reshape(-1, 4)
+        reduced_blocks = GroupedPruningMaskCreator.reduce_tensor(
+            blocked_tensor, 1, self._grouping_fn_name
+        )
+        return reduced_blocks.type(tensor.type())
+
+    def _map_mask_to_tensor(
+        self,
+        grouped_mask: Tensor,
+        original_tensor_shape: torch.Size,
+        tensor_idx: Optional[int] = None,
```

Comment:
`tensor_idx` doesn't get used at the moment. Is it included now for the future?

Reply:
this can be used for other implementing classes, not currently used for block masks